### PR TITLE
adding extract command to CLI

### PR DESCRIPTION
**Pull request by SimonGreenhill**
This PR adds a command to extract a subset of the data to a csv file for future analysis. 

```
dplace extract --dataset EA --tree dunn_et_al2011_utoaztecan output.csv
dplace extract --dataset EA --tree dunn_et_al2011_utoaztecan --variable EA001 output.csv
```

### CODE REVIEW

File: pydplace/commands.py [hunk 1]
```diff
@@ -110,3 +112,70 @@ def isclose(a, b, rel_tol=1e-09, abs_tol=0.0):
...
+    else:
+        variables = {v.id: v for v in ds.variables}
+    variable_list = sorted(variables.keys())
+
+    # collect data
+    glottocodes = [t.glottocode for t in tree.taxa]
+    societies = {s.id: s for s in ds.societies if s.glottocode in glottocodes}
+    
+    # prefilter data to remove things we don't want -> speed up the next step
+    data = [d for d in ds.data if d.soc_id in societies and d.var_id in variable_list]
```
Comment (xrotwang) [inline pydplace/commands.py]: If societies and/or variables may come from more than one dataset, we'd have to loop over all datasets here.

**Comment by xrotwang** [comment]
One could add another option `--codes`, and maybe output the results as GeoJSON, and voila: All matrilocal societies on a map! Look no hands, no web-app!

**Comment by xrotwang** [comment]
Oh, I just realized that `--tree` is used to select a bunch of societies, too. Hm, as long as tree/phylogeny IDs don't clash with dataset IDs, we could just allow any of these identifiers to be listed for `--societies`.

**Comment by SimonGreenhill** [comment]
Hmm. Yes, I wrote it for the use-case of extracting a set of data filtered by a tree. So what would a more general command look like?

Three arguments --society --variable --tree; all filter their respective parameters, so:

```
dplace extract --society x                                   # get all vars for society x
dplace extract --variable a                                 # extract all variable a
dplace extract --society x,y,z --variable a,b,c    # get vars a,b,c from societies x,y,z
dplace extract --tree x --variable a,b,c               # get vars a,b,c from societies on tree x (where x may be glottolog.global.trees
```

This gets complex fast, right? and there's some oddness i.e. all arguments except --tree take multiple values (multiple trees doesn't make sense as we can't join two trees into one without making analytic decisions). What's the minimal functionality this needs?

**Comment by xrotwang** [comment]
The way I see it, we want to filter datapoints, i.e. points in the society-variable matrix. The `--tree` argument is just a short-cut to filter societies, so maybe the arguments should rather be `--societies-from-tree` and `--societies-from-dataset` and maybe an additional `--societies`, which may accept a list of society IDs. I don't think it needs to be more complicated than that, right?

**Comment by SimonGreenhill** [comment]
Ok -- how's this? Followed most of your suggestions, but ended up making it fuller featured anyway. To make the command code tidier I've added:

1. an iter_data method of Repos, which returns a generator of the records, this can be filtered by society ids, variable ids, and dataset ids. All filtering is done in-line so it's quicker and less memory hungry than the post-hoc filtering that was being done earlier (e.g. [for r in ds.data if r...]). 

2. Adds a `societies` and `variables` dictionary to the top level repository object to make getting details of these easier.

Output is now in 'long' format rather than 'wide'  (i.e. one line per data record rather than one line per society). As this was quicker (I don't need to identify which variables are retained and sort to get the ordering right), but this does mean each row has lots of duplicate data (society name, lat, long, glottocode etc). Thoughts on this?

**Comment by xrotwang** [comment]
Looks pretty feature complete now. Long format is fine with me (no wonder, CLDF is long, too :) ) - and I think aggregating / grpuping is something
- each analysis software can do well
- each researcher should know how to do.

**Comment by kirbykat** [comment]
@xrotwang @SimonGreenhill If possible to keep the data download as a wide download, I think that would make it more accessible to, e.g., undergrad students/less code-savvy researchers. My two cents, it makes no difference to me, personally.

**Comment by xrotwang** [comment]
@kirbykat this pull request is not about the web app. I think people who would use this python/commandline API to extract data for further analysis would be able to transform long to wide easily.

**Comment by kirbykat** [comment]
@xrotwang @SimonGreenhill Ok, great. On a related note - I've had a couple of requests via the D-PLACE email for the "full" dataset. Any thoughts on the best response to these requests? Some people may not have the skills to use the python/commandline API to extract data...is there a "link" that would explain how to do this? I guess the alternative is to send the various input datasets directly to those requesting them.

**Comment by xrotwang** [comment]
@kirbykat The best response would be the link to this repository, I'd say. I think in a couple of weeks we should have some documentation on installation and usage of the python client API available.

**Pull request merged**